src/CashuWallet.ts

```diff
@@ -987,23 +1022,53 @@ class CashuWallet {
 	 * @param rs arrays of binding factors
 	 * @param secrets array of secrets
 	 * @param keyset mint keyset
+	 * @param verifyDLEQ require proof of same secret (DLEQ)
 	 * @returns array of serialized proofs
 	 */
 	private constructProofs(
 		promises: Array<SerializedBlindedSignature>,
 		rs: Array<bigint>,
 		secrets: Array<Uint8Array>,
-		keyset: MintKeys
+		keyset: MintKeys,
+		verifyDLEQ: boolean
 	): Array<Proof> {
-		return promises
-			.map((p: SerializedBlindedSignature, i: number) => {
-				const blindSignature = { id: p.id, amount: p.amount, C_: pointFromHex(p.C_) };
-				const r = rs[i];
-				const secret = secrets[i];
-				const A = pointFromHex(keyset.keys[p.amount]);
-				return constructProofFromPromise(blindSignature, r, secret, A);
-			})
-			.map((p: NUT11Proof) => serializeProof(p) as Proof);
+		return promises.map((p: SerializedBlindedSignature, i: number) => {
+			const dleq =
+				p.dleq == undefined
+					? undefined
+					: ({
+							s: hexToBytes(p.dleq.s),
+							e: hexToBytes(p.dleq.e),
+							r: rs[i]
+					} as DLEQ);
+			const blindSignature = {
+				id: p.id,
+				amount: p.amount,
+				C_: pointFromHex(p.C_),
+				dleq: dleq
+			};
+			const r = rs[i];
+			const secret = secrets[i];
+			const A = pointFromHex(keyset.keys[p.amount]);
+			const proof = constructProofFromPromise(blindSignature, r, secret, A);
+			if (verifyDLEQ) {
+				if (dleq == undefined) {
+					throw new Error('DLEQ verification required, but none found');
+				}
+				if (!verifyDLEQProof_reblind(secret, dleq, proof.C, A)) {
+					throw new Error('DLEQ verification failed');
+				}
+			}
```

Comment:
```suggestion
			if (verifyDLEQ) {
				if (dleq == undefined) {
					throw new Error('DLEQ verification required, but none found');
				}
				if (!verifyDLEQProof_reblind(secret, dleq, proof.C, A)) {
					throw new Error('DLEQ verification failed');
				}
			}
```

It might be dangerous to throw an error here, we might still want to keep the proofs, even if the verification failed. I have no good solution for this case though.

Reply:
> It might be dangerous to throw an error here

Agreed.

What if cashu-ts just constructs the dleq, then its up to the implementer to verify and handle failures?

The other idea I had was have some sort of flag either on the `Proof` or in the return value of `constructProofs` that indicates the verification failed, but that gets gross:

```typescript
/**
 * represents a single Cashu proof.
 */
export type Proof = {
	...

         /**
	 * DLEQ proof
	 */
	dleq?: SerializedDLEQ;
        /**
        * Is the DLEQ proof valid
        */
         dleqValid?: boolan;
};
````

Reply:
I really like @gudnuf 's idea

Reply:
What do you think @callebtc

Reply:
nostr-tools uses a Symbol to signal verified events: https://github.com/nbd-wtf/nostr-tools/blob/master/core.ts

Reply:
@Egge21M What changes from a normal boolean flag?

Reply:
Symbols are a bit "safer" than regular properties. The are not enumerable and have to be accessed very purposefully. Also, because they are not enumerable they will be excluded by default by methods like JSON.stringify